
<h1 align="center">  Restaurant and Coffee House Chains in TORONTO </h1>

<h2 align="center"> Applied Data Science Capstone by IBM/Coursera (Week-01) </h2>

<h4 align="center">  </h4>

<h4 align="center"> Author: Andrei Borodich </h4>

<h4 align="center"> Date: 27-FEB-2020 </h4>

<h4 align="center">  </h4>




The URL to the Jupyther notebook at GitHub <br>
https://github.com/and44bor/Coursera_Capstone/blob/master/RBI--week1.ipynb


The URL to the rendered html-page at GitHub <br>
https://htmlpreview.github.io/?https://github.com/and44bor/Coursera_Capstone/blob/master/RBI--week1.html


## Table of contents
* [Introduction: Business Problem](#introduction)
* [Data](#data)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion)


## Introduction: Business Problem <a name="introduction"></a>

In this project we explore the West End and East End of the *City of Toronto* in order to find attractive locations to open a few new food venues that are members of different franchises though running by the same management company. 
Specifically, this report will cover business planning issues pertinent to the fast-food chain companies .  The management companies like the **Restaurant Brands International Inc. (RBI)** show their  interest to this type of research analytics. Because it provides the grounded on data, scientific approach to estimate the optimal location for the particular type of their brands. 

The case description may look as it follows. 
The RBI which is a Canadian multinational fast-food holding company, owns  American fast-food restaurant chain  Burger King (and Popeyes as well) and Canadian coffee shop chain Tim Hortons.  Relaying on marketing data, the management promotes opening a few new franchisees in the "Low Venue Density" neighborhoods of Toronto that are predominantly located in the two boroughs, *Scarborough* (East End) and *Etobicoke* (West End) of the city.   
The task is to locate particular neighborhoods having the high affinity to the fast-food venues. Then, identify those that more suitable for opening a new **fast-food restaurant (Burger King)** and other for  opening a new **coffee shop (Tim Hortons)**. 

We will use the Foursquare location data to create datasets of neighborhood venues. 
We will explore a given borough and segment it with the pandas dataframe utilities and methods.   
The descriptive statistics applied to the datasets will help to find locations with high popularity of the **fast-food restaurants (FFRs)** and **coffee shops (CSs)**. The k-Means  Clustering  Method  will be run on the neighborhood--venue data and group neighborhoods based on the similarity of their venue features. We will provide a list of particular locations of the high affinity to the fast-food venues as well as discuss their suitability to open either a new FFR or CS.



## Data <a name="data"></a>

We begin with creating a representative dataset with the Toronto neighborhood geospatial data. 
The City of Toronto is divided into 6 districts (boroughs). 
They are Old Toronto (Downtown, Midtown and Islands), York, East York, North York, Scarborough and Etobicoke. 
In the previous work, "*Segmenting and Clustering Neighborhoods in Toronto*", we have created 2 dataframes with the boroughs' and neighborhoods' features and stored the datasets into 2 csv-files. 

One stored dataset contains the names of neighborhoods combined under the same postal codes. 
It was obtained by making calls to the reliable web-site and fetching the html-table from the Toronto web-page into a pandas dataframe.
We have used the `lxml`  Python library to do a web-site scraping.

Other stored dataset contains the geographical coordinates of combined neighborhoods. 
It was obtained by retrieving  CanadaPost's geocoded data from the postal codes.
We have used the `geocoder` Python library to get the geospatial coordinates of combined neighborhoods linked to the postal codes. 

With the 2 availables datasets we create a combined dataframe of the Toronto neighborhoods including features as name, borough label, latitude, longitude and postal code affiliation. 
From that dataframe we make the two reduced ones for two boroughs, *Scarborough* and *Etobicoke*. Those dataframes, `scarbo_data` and `etobi_data`, are output below.


In [83]:
scarbo_data.head(5)

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [84]:
etobi_data.head(5)

PostalCode    Borough                                       Neighborhood  \
0        M8V  Etobicoke       Humber Bay Shores, Mimico South, New Toronto   
1        M8W  Etobicoke                             Alderwood, Long Branch   
2        M8X  Etobicoke      The Kingsway, Montgomery Road, Old Mill North   
3        M8Y  Etobicoke  Humber Bay, King's Mill Park, Kingsway Park So...   
4        M8Z  Etobicoke  Kingsway Park South West, Mimico NW, The Queen...   

    Latitude  Longitude  
0  43.605647 -79.501321  
1  43.602414 -79.543484  
2  43.653654 -79.506944  
3  43.636258 -79.498509  
4  43.628841 -79.520999

There are **17 combined neighborhoods** in the **Scarborough** borough. 

There are **11 combined neighborhoods** in the **Etobicoke** borough. 

We also draw the map of each borough and add neigborhood locations onto the map. 
The maps were made with a rendering library `Folium`, which is a great visualization library. The Folium maps are easy to navigate and fast to respond zooming. 
The nominal geospatial coordinates (latitude and longitude) of Scarborough, ON and Etobicoke, ON were obtained with with the geolocator API of the geolocation service `Nominatim` from the GeoPy library.

Then, we proceed with creating datasets of the **neighborhood most popular venues**, one for Scarborough and other for Etobicoke.
We explore neighborhoods in a given borough utilizing the Foursquare API. By making calls we get the top popular venues, with the limit of 100,  of any categories within a radius of 1250 meters around every neighborhood. As we empirically determined it is an optimal radius for a bigger circular coverage but fewer venue overlap in case of the "low venue density" neighborhoods of Toronto. 
We collect results and extracts the venue features from the json-structure. 
We check the venues counted more than once, because of searching circle overlap, retain the proper venue IDs and discard their duplicates. 
Hence, we create two datasets of neighborhood venues for two boroughs, Scarborough and Etobicoke. 
Those dataframes, `scarbo_venues ` and `etobi_venues `, are output below. 
We visualize the venue locations on the borough's maps. 


In [85]:
scarbo_venues.head(5)

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Rouge, Malvern              43.806686              -79.194353   
1  Rouge, Malvern              43.806686              -79.194353   
2  Rouge, Malvern              43.806686              -79.194353   
3  Rouge, Malvern              43.806686              -79.194353   
4  Rouge, Malvern              43.806686              -79.194353   

                 Venue                  Venue id  Venue Latitude  \
0              Wendy's  579a91b3498e9bd833afa78a       43.802008   
1             Harvey's  4b6718c2f964a5203f3a2be3       43.800020   
2              Wendy's  4bb6b9446edc76b0d771311c       43.807448   
3       Caribbean Wave  4b914562f964a520d4ae33e3       43.798558   
4  Staples Morningside  4bcb612d3740b713f0606265       43.800285   

   Venue Longitude                 Venue Category  
0       -79.198080           Fast Food Restaurant  
1       -79.198307                     Restaurant  
2       -79.199056           Fast Food Restaurant  
3       -79.195777           Caribbean Restaurant  
4       -79.196607  Paper / Office Supplies Store

In [86]:
etobi_venues.head(5)

Neighborhood  Neighborhood Latitude  \
0  Humber Bay Shores, Mimico South, New Toronto              43.605647   
1  Humber Bay Shores, Mimico South, New Toronto              43.605647   
2  Humber Bay Shores, Mimico South, New Toronto              43.605647   
3  Humber Bay Shores, Mimico South, New Toronto              43.605647   
4  Humber Bay Shores, Mimico South, New Toronto              43.605647   

   Neighborhood Longitude                   Venue                  Venue id  \
0              -79.501321                    LCBO  4b119977f964a520488023e3   
1              -79.501321          Huevos Gourmet  5395d784498e085ff3c18198   
2              -79.501321          Sweet Olenka's  4c081867ffb8c9b6d7b66761   
3              -79.501321          Kitchen on 6th  571cdf3e38faaa76e74049d7   
4              -79.501321  Cellar Door Restaurant  53938d4c498ea919cfa0117f   

   Venue Latitude  Venue Longitude      Venue Category  
0       43.602281       -79.499302        Liquor Store  
1       43.601188       -79.503717  Mexican Restaurant  
2       43.601099       -79.500325        Dessert Shop  
3       43.601396       -79.504563      Breakfast Spot  
4       43.600221       -79.507638  Italian Restaurant

Then, we proceed with creating datasets of the **neighborhood particular category venues**, as FFRs and CSs, for Scarborough and Etobicoke.
Again, we utilize the Foursquare API. We get only FFRs and CSs, each category having the limit of 100,  within a radius of 1250 meters around every neighborhood. 
After extracting venue features from the json-structure and discarding the venue duplicates  we create four more datasets, 
Those dataframes, `df_FFRs `  \&  `df_CSs ` for Scarborough  and `df_FFRe `  \&  `df_CSe ` for Etobicoke are output below. 
We visualize the venue locations on the borough's maps. 

In [87]:
df_FFRs.head(5) 
# df_CSs.head(5) 

nei                  venue_id  venue_name  venue_lat  venue_lon  \
0    1  579a91b3498e9bd833afa78a     Wendy's  43.802008 -79.198080   
1    1  4bb6b9446edc76b0d771311c     Wendy's  43.807448 -79.199056   
2    1  4b89abadf964a520ae4a32e3     Quiznos  43.798773 -79.196556   
3    1  4c3660c80a71c9b6d30f3dc9  McDonald's  43.798261 -79.201449   
4    3  4b9023e9f964a5200e7833e3  McDonald's  43.768334 -79.188288   

                                          venue_addr  \
0       8129 Sheppard Avenue, Scarborough ON M1B 6A3   
1                                         Toronto ON   
2  875 Milner Ave (at Morningside Ave), Toronto O...   
3  785 Milner Avenue (Walmart), Scarborough ON M1...   
4  4435 Kingston Rd. (Lawrence Ave E), Scarboroug...   

                    cate_id             cate_name  
0  4bf58dd8d48988d16e941735  fast food restaurant  
1  4bf58dd8d48988d16e941735  fast food restaurant  
2  4bf58dd8d48988d16e941735  fast food restaurant  
3  4bf58dd8d48988d16e941735  fast food restaurant  
4  4bf58dd8d48988d16e941735  fast food restaurant

In [88]:
# df_FFRe.head(5) 
df_CSe.head(5)

nei                  venue_id                    venue_name  venue_lat  \
0    1  5086d512e4b0aafd9871629f                   Tim Hortons  43.598932   
1    1  4f0c86b7e4b0d563f7dd6e9c                   Coffee Time  43.602284   
2    1  4be6c47a2457a593101cad15  Big Guy's Little Coffee Shop  43.601359   
3    2  52abc3b011d279cc571a9e77                     Starbucks  43.594427   
4    2  55e4f8ee498e1696f68060e8                   Tim Hortons  43.592881   

   venue_lon                                         venue_addr  \
0 -79.511990  3089 Lakeshore Blvd (Lakeshore), Etobicoke ON ...   
1 -79.499857     2786 Lake Shore Blvd W (at 2nd St), Toronto ON   
2 -79.502480                  2861 Lakeshore Blvd W, Toronto ON   
3 -79.533384  3559 Lake Shore Blvd W (Thirty Third St), Toro...   
4 -79.539087     3719 Lakeshore Blvd West, Etobicoke ON M8W 1P8   

                    cate_id    cate_name  
0  4bf58dd8d48988d1e0931735  coffee shop  
1  4bf58dd8d48988d1e0931735  coffee shop  
2  4bf58dd8d48988d1e0931735  coffee shop  
3  4bf58dd8d48988d1e0931735  coffee shop  
4  4bf58dd8d48988d1e0931735  coffee shop

## END

# The Working Code

In [1]:
import numpy as np       # library to handle vector arrays
import pandas as pd      # library for data analysis

import json         # library to handle JSON files
import requests     # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

from sklearn.cluster import KMeans    # import k-means for clustering data

# use the inline backend to generate the plots within the browser
%matplotlib inline 
#
import matplotlib as mpl
import matplotlib.pyplot as plt
#
# mpl.style.use('ggplot')   # optional: for ggplot-like style
# print ('Matplotlib version: ', mpl.__version__) # optional: version of Matplotlib
#
import matplotlib.cm as cm
import matplotlib.colors as colors

# !conda install -c conda-forge folium --yes 
import folium     # map rendering library

import math
import re


In [2]:
pip install geopy    # run in separate cell

     |████████████████████████████████| 112kB 9.5MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [3]:
from geopy.geocoders import Nominatim
from geopy.distance import geodesic
from geopy.distance import lonlat, distance


In [ ]:
del [ scarbo_venues, etobi_venues ]

<a id='item1'></a>

### (1) Toronto neighborhoods: Locations Affiliated with Postal Codes

In order to segment the neighborhoods and explore them, we essentially need:  <br>
(1) the data that contains the boroughs and the neighborhoods that exist in each borough, <br>
(2) the latitude and longitude coordinates of each neighborhood. <br>


In [6]:
# Read csv-file into pandas DataFrame
# NOTE: DataFrame with 3 columns named as PostalCode, Borough, Neighborhood.
# NOTE: DataFrame represents 11 boroughs and 208 neighborhoods of the City of Toronto. 
#
df2 = pd.read_csv('toronto_code_borrow.csv')
df2

# df2.sort_values("PostalCode", axis = 0, ascending = True).reset_index(drop=True, inplace=True)

PostalCode      Borough                                       Neighborhood
0          M1B  Scarborough                                     Rouge, Malvern
1          M1C  Scarborough             Highland Creek, Rouge Hill, Port Union
2          M1E  Scarborough                  Guildwood, Morningside, West Hill
3          M1G  Scarborough                                             Woburn
4          M1H  Scarborough                                          Cedarbrae
..         ...          ...                                                ...
98         M9N         York                                             Weston
99         M9P    Etobicoke                                          Westmount
100        M9R    Etobicoke  Kingsview Village, Martin Grove Gardens, Richv...
101        M9V    Etobicoke  Albion Gardens, Beaumond Heights, Humbergate, ...
102        M9W    Etobicoke                                          Northwest

[103 rows x 3 columns]

### (2) Toronto neighborhoods: Geographical Coordinates

In [7]:
# Read csv-file into pandas DataFrame
# NOTE: DataFrame with 3 columns named as Postal Code, Latitude, Longitude.
# NOTE: DataFrame represents 103 postal-code-centers in the City of Toronto. 
#
dg0 = pd.read_csv('Geospatial_Coordinates.csv', header=0)
dg0

# dg0.sort_values("Postal Code", axis = 0, ascending = True).reset_index(drop=True, inplace=True)

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

### (3) Toronto neighborhoods: Combined Dataframe

In [8]:
# Re-arrange 2 dataframes to create the new one
#
df3 = pd.concat([ df2[['PostalCode', 'Borough', 'Neighborhood']], dg0 [['Latitude', 'Longitude']] ], axis=1)
df3


PostalCode      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
..         ...          ...   
98         M9N         York   
99         M9P    Etobicoke   
100        M9R    Etobicoke   
101        M9V    Etobicoke   
102        M9W    Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, Martin Grove Gardens, Richv...  43.688905 -79.554724  
101  Albion Gardens, Beaumond Heights, Humbergate, ...  43.739416 -79.588437  
102                                          Northwest  43.706748 -79.594054  

[103 rows x 5 columns]

### (4) Scarborough and its neighborhoods

We are going to segment and cluster the neighborhoods in **Scarborough**. <br> 
For that we create a new dataframe of the **Scarborough** data only. <br>
Then we visualize the **Scarborough** and the neighborhoods in this borough.

In [9]:
# create a DataFrame of Scarborough data only
#
scarbo_data = df3[ df3['Borough'] == 'Scarborough'].reset_index(drop=True)
scarbo_data


PostalCode      Borough                                       Neighborhood  \
0         M1B  Scarborough                                     Rouge, Malvern   
1         M1C  Scarborough             Highland Creek, Rouge Hill, Port Union   
2         M1E  Scarborough                  Guildwood, Morningside, West Hill   
3         M1G  Scarborough                                             Woburn   
4         M1H  Scarborough                                          Cedarbrae   
5         M1J  Scarborough                                Scarborough Village   
6         M1K  Scarborough        East Birchmount Park, Ionview, Kennedy Park   
7         M1L  Scarborough                    Clairlea, Golden Mile, Oakridge   
8         M1M  Scarborough    Cliffcrest, Cliffside, Scarborough Village West   
9         M1N  Scarborough                        Birch Cliff, Cliffside West   
10        M1P  Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...   
11        M1R  Scarborough                                  Maryvale, Wexford   
12        M1S  Scarborough                                          Agincourt   
13        M1T  Scarborough            Clarks Corners, Sullivan, Tam O'Shanter   
14        M1V  Scarborough  Agincourt North, L'Amoreaux East, Milliken, St...   
15        M1W  Scarborough                                    L'Amoreaux West   
16        M1X  Scarborough                                        Upper Rouge   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577  
15  43.799525 -79.318389  
16  43.836125 -79.205636

There are **17 combined neighborhoods** in the **Scarborough** borough. 

In [10]:

geolocator = Nominatim(user_agent="scarbo_explorer")
location = geolocator.geocode('SCARBOROUGH, ON')
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of SCARBOROUGH, ON are {}, {}.'.format(latitude, longitude))


The geograpical coordinate of SCARBOROUGH, ON are 43.773077, -79.257774.


In [11]:
# create a map of Scarborough (Toronto) using latitude and longitude values
map_scarbo = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers of neighborhoods to the map
for lat, lng, neighb in zip(scarbo_data['Latitude'], scarbo_data['Longitude'], scarbo_data['Neighborhood']):
    label = '{}'.format(neighb)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=20,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.2,
        parse_html=False).add_to(map_scarbo)

map_scarbo

### (5) Etobicoke and its neighborhoods

We are going to segment and cluster the neighborhoods in **Etobicoke**. <br> 
For that we create a new dataframe of the **Etobicoke** data only. <br>
Then we visualize the **Etobicoke** and the neighborhoods in this borough.

In [13]:
# create a DataFrame of Etobicoke data only
#
etobi_data = df3[ df3['Borough'] == 'Etobicoke'].reset_index(drop=True)
etobi_data


PostalCode    Borough                                       Neighborhood  \
0         M8V  Etobicoke       Humber Bay Shores, Mimico South, New Toronto   
1         M8W  Etobicoke                             Alderwood, Long Branch   
2         M8X  Etobicoke      The Kingsway, Montgomery Road, Old Mill North   
3         M8Y  Etobicoke  Humber Bay, King's Mill Park, Kingsway Park So...   
4         M8Z  Etobicoke  Kingsway Park South West, Mimico NW, The Queen...   
5         M9B  Etobicoke  Cloverdale, Islington, Martin Grove, Princess ...   
6         M9C  Etobicoke  Bloordale Gardens, Eringate, Markland Wood, Ol...   
7         M9P  Etobicoke                                          Westmount   
8         M9R  Etobicoke  Kingsview Village, Martin Grove Gardens, Richv...   
9         M9V  Etobicoke  Albion Gardens, Beaumond Heights, Humbergate, ...   
10        M9W  Etobicoke                                          Northwest   

     Latitude  Longitude  
0   43.605647 -79.501321  
1   43.602414 -79.543484  
2   43.653654 -79.506944  
3   43.636258 -79.498509  
4   43.628841 -79.520999  
5   43.650943 -79.554724  
6   43.643515 -79.577201  
7   43.696319 -79.532242  
8   43.688905 -79.554724  
9   43.739416 -79.588437  
10  43.706748 -79.594054

There are **11 combined neighborhoods** in the **Etobicoke** borough. 

In [14]:

geolocator = Nominatim(user_agent="etobi_explorer")
location = geolocator.geocode('ETOBICOKE, ON')
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of ETOBICOKE, ON are {}, {}.'.format(latitude, longitude))


The geograpical coordinate of ETOBICOKE, ON are 43.671459150000004, -79.55249206611668.


In [15]:
# create a map of Etobicoke (Toronto) using latitude and longitude values
map_etobi = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers of neighborhoods to the map
for lat, lng, neighb in zip(etobi_data['Latitude'], etobi_data['Longitude'], etobi_data['Neighborhood']):
    label = '{}'.format(neighb)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=20,
        popup=label,
        color='green',
        fill=True,
        fill_color='#71bf30',
        fill_opacity=0.2,
        parse_html=False).add_to(map_etobi)

map_etobi

### (6) Define Foursquare Client Credentials and Version

Now we start utilizing the **Foursquare API** to explore neighborhoods in a given borough.

In [16]:
# For security reasons, personal FourSquare ID and SECRET should be stored in a separate file. 
# Here, the code reads testword, CLIENT_ID, CLIENT_SECRET, testword from the specified file.
# Then, VERSION is assigned to the Foursquare API Version.
#
with open ("FourSquare--CLIENT.txt", "r") as myfile:
    lines = myfile.read().splitlines()
myfile.close()
testword=lines[0]
CLIENT_ID=lines[1]
CLIENT_SECRET=lines[2]
testword=lines[3]
#
VERSION = '20200131' # Foursquare API version
#
print( 'Client ID and Client Secret have been obtained.')
print( 'The Version is {}.'.format(VERSION))
# print('CLIENT_ID: ' + CLIENT_ID)
# print('CLIENT_SECRET:' + CLIENT_SECRET)

Client ID and Client Secret have been obtained.
The Version is 20200131.


In [17]:
# function:   "getNearbyVenues"
# return "nearby_venues", a dataframe of venues of all categories, all neighborhoods
#
# creates a "venues_list" checking the neighborhoods one by one  
# makes the GET requests to Foursquare; extracts "8 features" from json-structure


def getNearbyVenues(names, latitudes, longitudes, RAD, LIM):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        
        limit = LIM   # limit number of venues returned by Foursquare API
        radius = RAD  # radius (meters) from the selected location
        
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            limit)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['id'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                             'Neighborhood Latitude',
                             'Neighborhood Longitude',
                             'Venue',
                             'Venue id',
                             'Venue Latitude',
                             'Venue Longitude',
                             'Venue Category']
    
    return(nearby_venues)


In [42]:
# function:   "get_CCss"
# return "df_CCss", a dataframe of venues of a particular category (sub-category), all neighborhoods
#
# creates "CC_category" and "CCss" lists checking the neighborhoods one by one  
# makes the GET requests to Foursquare; extracts "8 features" from json-structure
#
# [ii, venue_id, venue_name, venue_latlon[0], venue_latlon[1], venue_address, category_id, category_name]  
# ['nei','venue_id','venue_name','venue_lat','venue_lon','venue_addr','cate_id','cate_name']


def get_categories(categories):
    return [(cate['name'], cate['id']) for cate in categories]

def format_address(location):
    address = ', '.join(location['formattedAddress'])
    address = address.replace(', Canada', '')
    return address

def get_venues_near_location(lat, lon, category, client_id, client_secret, version, radius, limit):
    url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&categoryId={}&radius={}&limit={}'.format(
        client_id, client_secret, version, lat, lon, category, radius, limit)
    try:
        results = requests.get(url).json()['response']['groups'][0]['items']
        venues = [(item['venue']['id'],
                   item['venue']['name'],
                   get_categories(item['venue']['categories']),
                   (item['venue']['location']['lat'], item['venue']['location']['lng']),
                   format_address(item['venue']['location']),
                   item['venue']['location']['distance']) for item in results]        
    except:
        venues = []
    return venues


def get_CCss(lats, lons, RAD, LIM, CC_code, CC_keyw):
    CC_category = []
    CCss = []
    
    # Category, CC_code is passing to get_venues_near_location   
    # Category, CC_keyw is used in category_name   

    fs_client_id=CLIENT_ID          # Foursquare
    fs_client_secret=CLIENT_SECRET  # Foursquare
    fs_version=VERSION              # Foursquare
    
    print('... gettings venues around given locations ...')
    ii = 0
    for lat, lon in zip(lats, lons):
        ii += 1
        # print( ii, ' lat & lon : ', lat, lon)
        venues = get_venues_near_location(lat, lon, CC_code, fs_client_id, fs_client_secret, fs_version, RAD, LIM)
        for venue in venues:
            venue_id = venue[0]
            venue_name = venue[1]
            venue_categories = venue[2]
            venue_latlon = venue[3]
            venue_address = venue[4]
            venue_distance = venue[5]
            
            for cc in venue_categories: 
                category_name = cc[0].lower()
                category_id = cc[1] 
                if CC_keyw in category_name: 
                    CC_category = [ii, venue_id, venue_name, venue_latlon[0], venue_latlon[1], venue_address, category_id, category_name]  
                    CCss.append(CC_category)
                    
    df_CCss = pd.DataFrame( CCss )
    df_CCss.columns = ['nei','venue_id','venue_name','venue_lat','venue_lon','venue_addr','cate_id','cate_name']
    
    print('... done ...')
    return df_CCss


In [43]:
# Function:         # EXTRA #   # TESTED #
# print by blocks to inspect the structure

def blo_CCss(CCss):
    blo = []
    for blo in CCss:
        print(blo)
    return


In [44]:
# Category, Coffee Shop:          CC_code = '4bf58dd8d48988d1e0931735'   
# Category, Coffee Shop:          CC_keyw = 'coffee shop'   
# Category, Fast Food Restaurant: CC_code = '4bf58dd8d48988d16e941735'   
# Category, Fast Food Restaurant: CC_keyw = 'fast food restaurant' 
#
# NOTE: under '4bf58dd8d48988d16e941735' detected also: chinese restaurant, noodle house, pizza place
# NOTE: under '4bf58dd8d48988d1e0931735' detected also: café, tea room

def info_CCss(): 
    print('Calling for the "Category" venues around given locations')
    CC_code = '4bf58dd8d48988d16e941735' 
    CC_keyw = 'fast food restaurant'
    print(' CC_code = {} ; CC_keyw = {} '.format(CC_code, CC_keyw) )
    CC_code = '4bf58dd8d48988d1e0931735' 
    CC_keyw = 'coffee shop'
    print(' CC_code = {} ; CC_keyw = {} '.format(CC_code, CC_keyw) )
    return


### (7) Explore Neighborhoods' Venues in SCARBOROUGH (Toronto)

We explore all neighborhoods in **Scarborough.** <br>
We are going to collect the top venues (limit is **100**) returned by Foursquare API, in every neighborhood in **Scarborough**, within a radius of **1250 meters.** <br>

In [20]:
# Create the pandas dataframe "scarbo_venues" with the venues in Scarborough 
#
scarbo_venues = getNearbyVenues(names=scarbo_data['Neighborhood'], 
                                latitudes=scarbo_data['Latitude'], 
                                longitudes=scarbo_data['Longitude'], RAD=1250, LIM=100 )


Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge


Hence, the code has generated data for **17 neighborhoods** in the **Scarborough** borough. 

In [21]:
scarbo_venues

# print( "shape: ", scarbo_venues.shape, '\n' )
# print( "info:", scarbo_venues.info(), '\n' ) 


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0     Rouge, Malvern              43.806686              -79.194353   
1     Rouge, Malvern              43.806686              -79.194353   
2     Rouge, Malvern              43.806686              -79.194353   
3     Rouge, Malvern              43.806686              -79.194353   
4     Rouge, Malvern              43.806686              -79.194353   
..               ...                    ...                     ...   
531  L'Amoreaux West              43.799525              -79.318389   
532  L'Amoreaux West              43.799525              -79.318389   
533  L'Amoreaux West              43.799525              -79.318389   
534  L'Amoreaux West              43.799525              -79.318389   
535      Upper Rouge              43.836125              -79.205636   

                        Venue                  Venue id  Venue Latitude  \
0                     Wendy's  579a91b3498e9bd833afa78a       43.802008   
1                    Harvey's  4b6718c2f964a5203f3a2be3       43.800020   
2                     Wendy's  4bb6b9446edc76b0d771311c       43.807448   
3              Caribbean Wave  4b914562f964a520d4ae33e3       43.798558   
4         Staples Morningside  4bcb612d3740b713f0606265       43.800285   
..                        ...                       ...             ...   
531              Petro-Canada  4bdf9084e75c0f47c5c2cb03       43.793777   
532  Northhill Meats And Deli  4df4f040fa76abc3d871d54f       43.798475   
533       Kidstown Water Park  4c4f0c8a92b6a593ccacf070       43.808227   
534                      LCBO  4cc34b311e596dcb5f89c967       43.793970   
535               Beige-ville  5233006511d296849550ec7a       43.845100   

     Venue Longitude                 Venue Category  
0         -79.198080           Fast Food Restaurant  
1         -79.198307                     Restaurant  
2         -79.199056           Fast Food Restaurant  
3         -79.195777           Caribbean Restaurant  
4         -79.196607  Paper / Office Supplies Store  
..               ...                            ...  
531       -79.330900                    Gas Station  
532       -79.333323                  Deli / Bodega  
533       -79.309434                           Pool  
534       -79.329729                   Liquor Store  
535       -79.208755                     Donut Shop  

[536 rows x 8 columns]

Hence, we have the dataset of **536 venues** in **17 neighborhoods** in the **Scarborough** borough. 

Let's inspect the two working datasets, **scarbo_data** and **scarbo_venues**. 

In [22]:
print( len( scarbo_data['Neighborhood']), ' neighborhoods are in SCARBOROUGH.'  )
print( len( scarbo_venues['Neighborhood'].unique()), ' neighborhoods have venue records.' )
# scarbo_venues['Neighborhood'].unique()

17  neighborhoods are in SCARBOROUGH.
17  neighborhoods have venue records.


As we can see, **17 neighborhoods** (of the total number of **17**) have venue records. <br> 
Even **"Upper Rouge"** location,  adjacent to the large park area, has venues in the **Foursquare** database. <br> So, every neighborhood has its share in the venue analysis. 

Let's check the **duplicated** venues due to search overlap.

In [23]:
print( len( scarbo_venues['Venue id'].unique()), ' unique venues ' )
print( len( scarbo_venues['Venue id']), ' total venues \n')
#
print( 'Dataframe =scarbo_venues= contains duplicated venues due to search overlap.')
print( 'Dataframe =scarbo_venues= will be updated by removing duplicated venues. \n')

527  unique venues 
536  total venues 

Dataframe =scarbo_venues= contains duplicated venues due to search overlap.
Dataframe =scarbo_venues= will be updated by removing duplicated venues. 



In [24]:
# Finding the duplicates
# Getting a list of indices where pandas boolean series is True

bss = scarbo_venues.duplicated(['Venue id'], keep=False)
# bss.index[bss]
arss = bss.index[bss].values
arss

array([ 61,  64,  73,  77,  82,  83,  84,  87, 104, 124, 125, 127, 128,
       130, 166, 233, 310, 339])

In [25]:
scarbo_venues.iloc[ [ 61,  64,  73,  77,  82,  83,  84,  87, 104, 124, 125, 127, 128, 130, 166, 233, 310, 339 ] ]


Neighborhood  Neighborhood Latitude  \
61                                              Woburn              43.770992   
64                                              Woburn              43.770992   
73                                              Woburn              43.770992   
77                                              Woburn              43.770992   
82                                              Woburn              43.770992   
83                                              Woburn              43.770992   
84                                              Woburn              43.770992   
87                                           Cedarbrae              43.773136   
104                                          Cedarbrae              43.773136   
124                                          Cedarbrae              43.773136   
125                                          Cedarbrae              43.773136   
127                                          Cedarbrae              43.773136   
128                                          Cedarbrae              43.773136   
130                                          Cedarbrae              43.773136   
166        East Birchmount Park, Ionview, Kennedy Park              43.727929   
233    Cliffcrest, Cliffside, Scarborough Village West              43.716316   
310  Dorset Park, Scarborough Town Centre, Wexford ...              43.757410   
339                                  Maryvale, Wexford              43.750072   

     Neighborhood Longitude                             Venue  \
61               -79.216917  The Real McCoy Burgers And Pizza   
64               -79.216917                    Govardhan Thal   
73               -79.216917                     Bombay Palace   
77               -79.216917                     Fabian's Cafe   
82               -79.216917                             Mac's   
83               -79.216917                 Rexall PharmaPlus   
84               -79.216917                        Pizza Nova   
87               -79.239476  The Real McCoy Burgers And Pizza   
104              -79.239476                    Govardhan Thal   
124              -79.239476                 Rexall PharmaPlus   
125              -79.239476                             Mac's   
127              -79.239476                        Pizza Nova   
128              -79.239476                     Fabian's Cafe   
130              -79.239476                     Bombay Palace   
166              -79.262029                               KFC   
233              -79.239476                               KFC   
310              -79.273304                         Keilhauer   
339              -79.295849                         Keilhauer   

                     Venue id  Venue Latitude  Venue Longitude  \
61   4bf470256a31d13a9ec7952e       43.774081       -79.230496   
64   4bd48a734e32d13a4254bf80       43.767496       -79.228476   
73   4c632edeedd320a19ad5ae29       43.767057       -79.227682   
77   4e0a53e3d164028d170dc1b6       43.768158       -79.229159   
82   4bf471a094af2d7f299f3a72       43.773949       -79.230272   
83   4cdededf70bba1cdbe1a63c4       43.775313       -79.231020   
84   4c097c1bbbc676b0674a48d5       43.776973       -79.229830   
87   4bf470256a31d13a9ec7952e       43.774081       -79.230496   
104  4bd48a734e32d13a4254bf80       43.767496       -79.228476   
124  4cdededf70bba1cdbe1a63c4       43.775313       -79.231020   
125  4bf471a094af2d7f299f3a72       43.773949       -79.230272   
127  4c097c1bbbc676b0674a48d5       43.776973       -79.229830   
128  4e0a53e3d164028d170dc1b6       43.768158       -79.229159   
130  4c632edeedd320a19ad5ae29       43.767057       -79.227682   
166  519d731e498e937ddd96d09b       43.723232       -79.249682   
233  519d731e498e937ddd96d09b       43.723232       -79.249682   
310  4e08bdd152b165b7d7b5f39d       43.753870       -79.287940   
339  4e08bdd152b165b7d7b5f39d       43.753870       -79.287940   

           

In [26]:
# add markers of duplicated venues to the existing map

for ii in arss:
    lat = scarbo_venues['Venue Latitude'].iloc[ii]
    lng = scarbo_venues['Venue Longitude'].iloc[ii]
    label = '{}'.format(scarbo_venues['Venue'].iloc[ii])
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=8,
        popup=label,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.2,
        parse_html=False).add_to(map_scarbo)

map_scarbo

There are 9 venues counted twice. They are listed in the previous output and shown on the map with red circles. <br> 
After visual inspection of the neighborhood boundaries on the map, <br> 
we retain the venues with `iloc[ 64, 73, 77, 87, 124, 125, 127, 233, 339 ]` <br>
and discard the venues with `iloc[ 61, 82, 83, 84, 104, 128, 130, 166, 310 ]`.

In [27]:
scarbo_venues.drop( [scarbo_venues.index[61], 
                     scarbo_venues.index[82], 
                     scarbo_venues.index[83], 
                     scarbo_venues.index[84], 
                     scarbo_venues.index[104], 
                     scarbo_venues.index[128], 
                     scarbo_venues.index[130], 
                     scarbo_venues.index[166], 
                     scarbo_venues.index[310] ], inplace=True)

scarbo_venues.reset_index(drop=True)

# scarbo_venues
print( "shape: ", scarbo_venues.shape, '\n' )


shape:  (527, 8) 



Thus, we have updated the dataframe **scarbo_venues**.

In [28]:
# Write pandas DataFrame to the csv-file
scarbo_venues.to_csv('scarbo_venues_winter.csv', sep=',', index=False)


In [ ]:
# Remove a dataframe from RAM
del [ scarbo_venues ]

# Read csv-file into pandas DataFrame
scarbo_venues = pd.read_csv('scarbo_venues_winter.csv', header=0)


In [29]:
scarbo_venues


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0     Rouge, Malvern              43.806686              -79.194353   
1     Rouge, Malvern              43.806686              -79.194353   
2     Rouge, Malvern              43.806686              -79.194353   
3     Rouge, Malvern              43.806686              -79.194353   
4     Rouge, Malvern              43.806686              -79.194353   
..               ...                    ...                     ...   
522  L'Amoreaux West              43.799525              -79.318389   
523  L'Amoreaux West              43.799525              -79.318389   
524  L'Amoreaux West              43.799525              -79.318389   
525  L'Amoreaux West              43.799525              -79.318389   
526      Upper Rouge              43.836125              -79.205636   

                        Venue                  Venue id  Venue Latitude  \
0                     Wendy's  579a91b3498e9bd833afa78a       43.802008   
1                    Harvey's  4b6718c2f964a5203f3a2be3       43.800020   
2                     Wendy's  4bb6b9446edc76b0d771311c       43.807448   
3              Caribbean Wave  4b914562f964a520d4ae33e3       43.798558   
4         Staples Morningside  4bcb612d3740b713f0606265       43.800285   
..                        ...                       ...             ...   
522              Petro-Canada  4bdf9084e75c0f47c5c2cb03       43.793777   
523  Northhill Meats And Deli  4df4f040fa76abc3d871d54f       43.798475   
524       Kidstown Water Park  4c4f0c8a92b6a593ccacf070       43.808227   
525                      LCBO  4cc34b311e596dcb5f89c967       43.793970   
526               Beige-ville  5233006511d296849550ec7a       43.845100   

     Venue Longitude                 Venue Category  
0         -79.198080           Fast Food Restaurant  
1         -79.198307                     Restaurant  
2         -79.199056           Fast Food Restaurant  
3         -79.195777           Caribbean Restaurant  
4         -79.196607  Paper / Office Supplies Store  
..               ...                            ...  
522       -79.330900                    Gas Station  
523       -79.333323                  Deli / Bodega  
524       -79.309434                           Pool  
525       -79.329729                   Liquor Store  
526       -79.208755                     Donut Shop  

[527 rows x 8 columns]

### (8) Explore Neighborhoods' Venues in ETOBICOKE (Toronto)

We explore all neighborhoods in **Etobicoke.** <br>
We are going to collect the top venues (limit is **100**) returned by Foursquare API, in every neighborhood in **Etobicoke**, within a radius of **1250 meters.** <br>

In [32]:
# Create the pandas dataframe "scarbo_venues" with the venues in Etobicoke 
#
etobi_venues = getNearbyVenues(names=etobi_data['Neighborhood'], 
                                latitudes=etobi_data['Latitude'], 
                                longitudes=etobi_data['Longitude'], RAD=1250, LIM=100 )


Humber Bay Shores, Mimico South, New Toronto
Alderwood, Long Branch
The Kingsway, Montgomery Road, Old Mill North
Humber Bay, King's Mill Park, Kingsway Park South East, Mimico NE, Old Mill South, The Queensway East, Royal York South East, Sunnylea
Kingsway Park South West, Mimico NW, The Queensway West, Royal York South West, South of Bloor
Cloverdale, Islington, Martin Grove, Princess Gardens, West Deane Park
Bloordale Gardens, Eringate, Markland Wood, Old Burnhamthorpe
Westmount
Kingsview Village, Martin Grove Gardens, Richview Gardens, St. Phillips
Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown
Northwest


Hence, the code has generated data for **11 neighborhoods** in the **Etobicoke** borough. 

In [33]:
etobi_venues

# print( "shape: ", etobi_venues.shape, '\n' )
# print( "info:", etobi_venues.info(), '\n' ) 


Neighborhood  Neighborhood Latitude  \
0    Humber Bay Shores, Mimico South, New Toronto              43.605647   
1    Humber Bay Shores, Mimico South, New Toronto              43.605647   
2    Humber Bay Shores, Mimico South, New Toronto              43.605647   
3    Humber Bay Shores, Mimico South, New Toronto              43.605647   
4    Humber Bay Shores, Mimico South, New Toronto              43.605647   
..                                            ...                    ...   
365                                     Northwest              43.706748   
366                                     Northwest              43.706748   
367                                     Northwest              43.706748   
368                                     Northwest              43.706748   
369                                     Northwest              43.706748   

     Neighborhood Longitude                                           Venue  \
0                -79.501321                                            LCBO   
1                -79.501321                                  Huevos Gourmet   
2                -79.501321                                  Sweet Olenka's   
3                -79.501321                                  Kitchen on 6th   
4                -79.501321                          Cellar Door Restaurant   
..                      ...                                             ...   
365              -79.594054                         Woodbine Hotel & Suites   
366              -79.594054  Toronto Pearson International Airport Pet Park   
367              -79.594054                           Enterprise Rent-A-Car   
368              -79.594054                                   Fitness Two 4   
369              -79.594054                         Post Parade Dining Room   

                     Venue id  Venue Latitude  Venue Longitude  \
0    4b119977f964a520488023e3       43.602281       -79.499302   
1    5395d784498e085ff3c18198       43.601188       -79.503717   
2    4c081867ffb8c9b6d7b66761       43.601099       -79.500325   
3    571cdf3e38faaa76e74049d7       43.601396       -79.504563   
4    53938d4c498ea919cfa0117f       43.600221       -79.507638   
..                        ...             ...              ...   
365  4fa3ca28e4b05ddd65da97ae       43.712624       -79.592664   
366  53c181eb498eb5bf28f6ab80       43.704901       -79.604441   
367  4d764ad4f5e4b1f7b07e0187       43.715260       -79.589320   
368  4dadba464df0522cc5608a66       43.714482       -79.582903   
369  5174132ae4b0d069bec3e724       43.715040       -79.604529   

          Venue Category  
0           Liquor Store  
1     Mexican Restaurant  
2           Dessert Shop  
3         Breakfast Spot  
4     Italian Restaurant  
..                   ...  
365                Hotel  
366              Dog Run  
367  Rental Car Location  
368                  Gym  
369           Restaurant  

[370 rows x 8 columns]

Hence, we have the dataset of **370 venues** in **11 neighborhoods** in the **Etobicoke** borough. 

Let's inspect the two working datasets, **etobi_data** and **etobi_venues**. 

In [34]:
print( len( etobi_data['Neighborhood']), ' neighborhoods are in ETOBICOKE.'  )
print( len( etobi_venues['Neighborhood'].unique()), ' neighborhoods have venue records.' )
# etobi_venues['Neighborhood'].unique()

11  neighborhoods are in ETOBICOKE.
11  neighborhoods have venue records.


As we can see, **11 neighborhoods** (of the total number of **11**) have venue records. <br> 
So, every neighborhood has its share in the venue analysis. 

Let's check the **duplicated** venues due to search overlap.

In [35]:
print( len( etobi_venues['Venue id'].unique()), ' unique venues ' )
print( len( etobi_venues['Venue id']), ' total venues \n')
#
print( 'Dataframe =etobi_venues= contains duplicated venues due to search overlap.')
print( 'Dataframe =etobi_venues= will be updated by removing duplicated venues. \n')

363  unique venues 
370  total venues 

Dataframe =etobi_venues= contains duplicated venues due to search overlap.
Dataframe =etobi_venues= will be updated by removing duplicated venues. 



In [36]:
# Finding the duplicates
# Getting a list of indices where pandas boolean series is True

bee = etobi_venues.duplicated(['Venue id'], keep=False)
# bee.index[bee]
aree = bee.index[bee].values
aree

array([228, 244, 285, 286, 287, 299, 302, 304, 315, 316, 317, 326, 332,
       334])

In [37]:
etobi_venues.iloc[ [228, 244, 285, 286, 287, 299, 302, 304, 315, 316, 317, 326, 332, 334] ]

Neighborhood  Neighborhood Latitude  \
228  Cloverdale, Islington, Martin Grove, Princess ...              43.650943   
244  Bloordale Gardens, Eringate, Markland Wood, Ol...              43.643515   
285                                          Westmount              43.696319   
286                                          Westmount              43.696319   
287                                          Westmount              43.696319   
299                                          Westmount              43.696319   
302                                          Westmount              43.696319   
304                                          Westmount              43.696319   
315  Kingsview Village, Martin Grove Gardens, Richv...              43.688905   
316  Kingsview Village, Martin Grove Gardens, Richv...              43.688905   
317  Kingsview Village, Martin Grove Gardens, Richv...              43.688905   
326  Kingsview Village, Martin Grove Gardens, Richv...              43.688905   
332  Kingsview Village, Martin Grove Gardens, Richv...              43.688905   
334  Kingsview Village, Martin Grove Gardens, Richv...              43.688905   

     Neighborhood Longitude                 Venue                  Venue id  \
228              -79.554724           Tim Hortons  50d4866ee4b0c1554d94fef7   
244              -79.577201           Tim Hortons  50d4866ee4b0c1554d94fef7   
285              -79.532242           Tim Hortons  4ede72ca77c8274e01ddd28b   
286              -79.532242             No Frills  50a26679e4b0ffcdfe90d654   
287              -79.532242    Tim Hortons / Esso  53502b6f498ed97618d62ba8   
299              -79.532242             241 Pizza  4bd3cec4046076b002037771   
302              -79.532242  TTC 37 Islington Bus  51d62d13498ee12c4f75b852   
304              -79.532242                   GNC  4d9f62be0ea841bd6991dbd7   
315              -79.554724    Tim Hortons / Esso  53502b6f498ed97618d62ba8   
316              -79.554724           Tim Hortons  4ede72ca77c8274e01ddd28b   
317              -79.554724             No Frills  50a26679e4b0ffcdfe90d654   
326              -79.554724                   GNC  4d9f62be0ea841bd6991dbd7   
332              -79.554724  TTC 37 Islington Bus  51d62d13498ee12c4f75b852   
334              -79.554724             241 Pizza  4bd3cec4046076b002037771   

     Venue Latitude  Venue Longitude   Venue Category  
228       43.644705       -79.567659      Coffee Shop  
244       43.644705       -79.567659      Coffee Shop  
285       43.696170       -79.544922      Coffee Shop  
286       43.695887       -79.544835    Grocery Store  
287       43.696703       -79.546227      Coffee Shop  
299       43.696775       -79.544214      Pizza Place  
302       43.697042       -79.546385         Bus Line  
304       43.691625       -79.546012  Supplement Shop  
315       43.696703       -79.546227      Coffee Shop  
316       43.696170       -79.544922      Coffee Shop  
317       43.695887       -79.544835    Grocery Store  
326       43.691625       -79.546012  Supplement Shop  
332       43.697042       -79.546385         Bus Line  
334       43.696775       -79.544214      Pizza Place

In [38]:
# add markers of duplicated venues to the existing map

for ii in aree:
    lat = etobi_venues['Venue Latitude'].iloc[ii]
    lng = etobi_venues['Venue Longitude'].iloc[ii]
    label = '{}'.format(etobi_venues['Venue'].iloc[ii])
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=8,
        popup=label,
        color='red',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.2,
        parse_html=False).add_to(map_etobi)

map_etobi

There are 7 venues counted twice. They are listed in the previous output and on the map with red circles. <br> 
After visual inspection of the neighborhood boundaries on the map, <br> 
we retain the venues   with `iloc[ 244, 285, 286, 287, 299, 302, 326 ]` <br>
and discard the venues with `iloc[ 228, 304, 315, 316, 317, 332, 334 ]`.

In [39]:
etobi_venues.drop( [etobi_venues.index[228], 
                    etobi_venues.index[304], 
                    etobi_venues.index[315], 
                    etobi_venues.index[316], 
                    etobi_venues.index[317], 
                    etobi_venues.index[332], 
                    etobi_venues.index[334] ], inplace=True)

etobi_venues.reset_index(drop=True)

# scarbo_venues
print( "shape: ", etobi_venues.shape, '\n' )


shape:  (363, 8) 



Thus, we have updated the dataframe **etobi_venues**.

In [40]:
# Write pandas DataFrame to the csv-file
etobi_venues.to_csv('etobi_venues_winter.csv', sep=',', index=False)


In [ ]:
# Remove a dataframe from RAM
del [ etobi_venues ]

# Read csv-file into pandas DataFrame
etobi_venues = pd.read_csv('etobi_venues_winter.csv', header=0)


In [41]:
etobi_venues


Neighborhood  Neighborhood Latitude  \
0    Humber Bay Shores, Mimico South, New Toronto              43.605647   
1    Humber Bay Shores, Mimico South, New Toronto              43.605647   
2    Humber Bay Shores, Mimico South, New Toronto              43.605647   
3    Humber Bay Shores, Mimico South, New Toronto              43.605647   
4    Humber Bay Shores, Mimico South, New Toronto              43.605647   
..                                            ...                    ...   
358                                     Northwest              43.706748   
359                                     Northwest              43.706748   
360                                     Northwest              43.706748   
361                                     Northwest              43.706748   
362                                     Northwest              43.706748   

     Neighborhood Longitude                                           Venue  \
0                -79.501321                                            LCBO   
1                -79.501321                                  Huevos Gourmet   
2                -79.501321                                  Sweet Olenka's   
3                -79.501321                                  Kitchen on 6th   
4                -79.501321                          Cellar Door Restaurant   
..                      ...                                             ...   
358              -79.594054                         Woodbine Hotel & Suites   
359              -79.594054  Toronto Pearson International Airport Pet Park   
360              -79.594054                           Enterprise Rent-A-Car   
361              -79.594054                                   Fitness Two 4   
362              -79.594054                         Post Parade Dining Room   

                     Venue id  Venue Latitude  Venue Longitude  \
0    4b119977f964a520488023e3       43.602281       -79.499302   
1    5395d784498e085ff3c18198       43.601188       -79.503717   
2    4c081867ffb8c9b6d7b66761       43.601099       -79.500325   
3    571cdf3e38faaa76e74049d7       43.601396       -79.504563   
4    53938d4c498ea919cfa0117f       43.600221       -79.507638   
..                        ...             ...              ...   
358  4fa3ca28e4b05ddd65da97ae       43.712624       -79.592664   
359  53c181eb498eb5bf28f6ab80       43.704901       -79.604441   
360  4d764ad4f5e4b1f7b07e0187       43.715260       -79.589320   
361  4dadba464df0522cc5608a66       43.714482       -79.582903   
362  5174132ae4b0d069bec3e724       43.715040       -79.604529   

          Venue Category  
0           Liquor Store  
1     Mexican Restaurant  
2           Dessert Shop  
3         Breakfast Spot  
4     Italian Restaurant  
..                   ...  
358                Hotel  
359              Dog Run  
360  Rental Car Location  
361                  Gym  
362           Restaurant  

[363 rows x 8 columns]

## (9) Calling for FFRs and CSs in SCARBOROUGH (Toronto)

In [45]:

print('***SCARBOROUGH***')
lats=scarbo_data['Latitude']
lons=scarbo_data['Longitude']
RAD = 1250
LIM = 100
info_CCss()
#
CC_keyw = 'fast food restaurant'
CC_code = '4bf58dd8d48988d16e941735'
df_FFRs = get_CCss(lats, lons, RAD, LIM, CC_code, CC_keyw)
#    
CC_keyw = 'coffee shop' 
CC_code = '4bf58dd8d48988d1e0931735'
df_CSs = get_CCss(lats, lons, RAD, LIM, CC_code, CC_keyw)
#   
print('Total number of FFRs: ', len(df_FFRs.index))
print('Total number of CSs: ',  len(df_CSs.index))


***SCARBOROUGH***
Calling for the "Category" venues around given locations
 CC_code = 4bf58dd8d48988d16e941735 ; CC_keyw = fast food restaurant 
 CC_code = 4bf58dd8d48988d1e0931735 ; CC_keyw = coffee shop 
... gettings venues around given locations ...
... done ...
... gettings venues around given locations ...
... done ...
Total number of FFRs:  62
Total number of CSs:  46


In [49]:
df_FFRs.head(5) ;
df_CSs.head(5) ;

In [60]:
# Finding the duplicates
# Getting a list of indices where pandas boolean series is True

bss = df_FFRs.duplicated(['venue_id'], keep=False)
arss = bss.index[bss].values
print('Duplicates in FFRe: ', arss )
del [bss, arss]

bss = df_CSs.duplicated(['venue_id'], keep=False)
arss = bss.index[bss].values
print('Duplicates in CSe: ', arss )
del [bss, arss]


Duplicates in FFRe:  [24 33]
Duplicates in CSe:  []


In [57]:
df_FFRs.iloc[ [24, 33] ]

# df_CSs.iloc[ [] ]

nei                  venue_id venue_name  venue_lat  venue_lon  \
24    7  519d731e498e937ddd96d09b        KFC  43.723232 -79.249682   
33    9  519d731e498e937ddd96d09b        KFC  43.723232 -79.249682   

                                venue_addr                   cate_id  \
24  2383 Kingston Road, Toronto ON M1N 4E7  4bf58dd8d48988d16e941735   
33  2383 Kingston Road, Toronto ON M1N 4E7  4bf58dd8d48988d16e941735   

               cate_name  
24  fast food restaurant  
33  fast food restaurant

In [62]:
# Discard the duplicates and update the dataframes
# Dataframes: df_FFRs and df_CSs

df_FFRs.drop( [ df_FFRs.index[33] ], inplace=True)
df_FFRs.reset_index(drop=True)
print( "shape: ", df_FFRs.shape, '\n' )

# df_CSs.drop( [  ], inplace=True)
# df_CSs.reset_index(drop=True)
# print( "shape: ", df_CSs.shape, '\n' )


shape:  (61, 8) 



In [66]:
# Write pandas DataFrame to the csv-file
df_FFRs.to_csv('scarbo_FFRs_winter.csv', sep=',', index=False)
df_CSs.to_csv('scarbo_CSs_winter.csv', sep=',', index=False)


In [68]:
# Remove a dataframe from RAM
del [ df_FFRs, df_CSs ]

# Read csv-file into pandas DataFrame
df_FFRs = pd.read_csv('scarbo_FFRs_winter.csv', header=0)
df_CSs = pd.read_csv('scarbo_CSs_winter.csv', header=0) 


In [ ]:
df_FFRs.head(5) 
df_CSs.head(5) 

In [79]:
# print( df_FFRs.query( 'venue_name == "Burger King"' ).index )
# print( df_CSs.query( 'venue_name == "Tim Hortons"' ).index )
print( '' )
print( df_FFRs[ df_FFRs['venue_name'].str.contains('Burger King')].index )
print( df_CSs[  df_CSs['venue_name'].str.contains('Tim Hortons')].index ) 


Int64Index([41, 45], dtype='int64')
Int64Index([ 1,  2,  4,  8, 10, 11, 14, 16, 17, 18, 22, 23, 27, 29, 30, 31, 33,
            35, 36, 37, 38, 39, 42, 44, 45],
           dtype='int64')


In [81]:
# The geograpical coordinate of SCARBOROUGH, ON
[latitude, longitude ] = [ 43.773077, -79.257774]

# create a map of FFRs & CSs in Scarborough (Toronto) using latitude and longitude values
map_scarbo = folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers of neighborhoods to the map
for lat, lng, num in zip(df_FFRs['venue_lat'], df_FFRs['venue_lon'], df_FFRs.index):
    label = '{}'.format( df_FFRs.iloc[num].venue_name )
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color = 'brown' if (df_FFRs.iloc[num].venue_name == 'Burger King') else 'blue',
        fill=True,
        #fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_scarbo)
#
for lat, lng, num in zip(df_CSs['venue_lat'], df_CSs['venue_lon'], df_CSs.index):
    label = '{}'.format( df_CSs.iloc[num].venue_name )
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color = 'red' if (df_CSs.iloc[num].venue_name == 'Tim Hortons') else 'green',
        fill=True,
        #fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_scarbo)

map_scarbo

## (10) Calling for FFRs and CSs in ETOBICOKE (Toronto)

In [46]:

print('***ETOBICOKE***')
lats=etobi_data['Latitude']
lons=etobi_data['Longitude']
RAD = 1250
LIM = 100
info_CCss()
#
CC_keyw = 'fast food restaurant'
CC_code = '4bf58dd8d48988d16e941735'
df_FFRe = get_CCss(lats, lons, RAD, LIM, CC_code, CC_keyw)
#
CC_keyw = 'coffee shop' 
CC_code = '4bf58dd8d48988d1e0931735'
df_CSe = get_CCss(lats, lons, RAD, LIM, CC_code, CC_keyw)
#  
print('Total number of FFRe: ', len(df_FFRe.index))
print('Total number of CSe: ',  len(df_CSe.index))


***ETOBICOKE***
Calling for the "Category" venues around given locations
 CC_code = 4bf58dd8d48988d16e941735 ; CC_keyw = fast food restaurant 
 CC_code = 4bf58dd8d48988d1e0931735 ; CC_keyw = coffee shop 
... gettings venues around given locations ...
... done ...
... gettings venues around given locations ...
... done ...
Total number of FFRe:  21
Total number of CSe:  41


In [54]:
df_FFRe.head(5) ;
df_CSe.head(5) ;

In [63]:
# Finding the duplicates
# Getting a list of indices where pandas boolean series is True

bee = df_FFRe.duplicated(['venue_id'], keep=False)
aree = bee.index[bee].values
print('Duplicates in FFRs: ', aree )
del [bee, aree]

bee = df_CSe.duplicated(['venue_id'], keep=False)
aree = bee.index[bee].values
print('Duplicates in CSs: ', aree )
del [bee, aree]


Duplicates in FFRs:  []
Duplicates in CSs:  [23 26 29 30 31 32]


In [64]:
# df_FFRe.iloc[ [  ] ]

df_CSe.iloc[ [ 23, 26, 29, 30, 31, 32 ] ]

nei                  venue_id          venue_name  venue_lat  venue_lon  \
23    6  50d4866ee4b0c1554d94fef7         Tim Hortons  43.644705 -79.567659   
26    7  50d4866ee4b0c1554d94fef7         Tim Hortons  43.644705 -79.567659   
29    8  4ede72ca77c8274e01ddd28b         Tim Hortons  43.696170 -79.544922   
30    8  53502b6f498ed97618d62ba8  Tim Hortons / Esso  43.696703 -79.546227   
31    9  53502b6f498ed97618d62ba8  Tim Hortons / Esso  43.696703 -79.546227   
32    9  4ede72ca77c8274e01ddd28b         Tim Hortons  43.696170 -79.544922   

                                           venue_addr  \
23  555 Burnhamthorpe Rd (at The West Mall), Etobi...   
26  555 Burnhamthorpe Rd (at The West Mall), Etobi...   
29  245 Dixon Rd (Islington Avenue), Etobicoke ON ...   
30                 261 Dixon Rd, Etobicoke ON M9P 2M4   
31                 261 Dixon Rd, Etobicoke ON M9P 2M4   
32  245 Dixon Rd (Islington Avenue), Etobicoke ON ...   

                     cate_id    cate_name  
23  4bf58dd8d48988d1e0931735  coffee shop  
26  4bf58dd8d48988d1e0931735  coffee shop  
29  4bf58dd8d48988d1e0931735  coffee shop  
30  4bf58dd8d48988d1e0931735  coffee shop  
31  4bf58dd8d48988d1e0931735  coffee shop  
32  4bf58dd8d48988d1e0931735  coffee shop

In [65]:
# Discard the duplicates and update the dataframes
# Dataframes: df_FFRe and df_CSe

# df_FFRe.drop( [  ], inplace=True)
# df_FFRe.reset_index(drop=True)
# print( "shape: ", df_FFRe.shape, '\n' )

df_CSe.drop( [df_CSe.index[26], df_CSe.index[31], df_CSe.index[32] ], inplace=True)
df_CSe.reset_index(drop=True)
print( "shape: ", df_CSe.shape, '\n' )


shape:  (38, 8) 



In [67]:
# Write pandas DataFrame to the csv-file
df_FFRe.to_csv('etobi_FFRe_winter.csv', sep=',', index=False)
df_CSe.to_csv('etobi_CSe_winter.csv', sep=',', index=False)


In [72]:
# Remove a dataframe from RAM
del [ df_FFRe , df_CSe ]

# Read csv-file into pandas DataFrame
df_FFRe = pd.read_csv('etobi_FFRe_winter.csv', header=0)
df_CSe = pd.read_csv('etobi_CSe_winter.csv', header=0) 


In [ ]:
df_FFRe.head(5) 
df_CSe.head(5)

In [80]:
# print( df_FFRe.query( 'venue_name == "Burger King"' ).index )
# print( df_CSe.query( 'venue_name == "Tim Hortons"' ).index )
print( '' )
print( df_FFRe[ df_FFRe['venue_name'].str.contains('Burger King')].index )
print( df_CSe[  df_CSe['venue_name'].str.contains('Tim Hortons')].index ) 
print( '' )


Int64Index([], dtype='int64')
Int64Index([0, 4, 5, 6, 12, 14, 15, 17, 18, 19, 20, 21, 23, 28, 29, 31, 34, 36,
            37],
           dtype='int64')



In [82]:
# The geograpical coordinate of ETOBICOKE, ON
[latitude, longitude ] = [ 43.671459, -79.552492]

# create a map of FFRe & CSe in Etobicoke (Toronto) using latitude and longitude values
map_etobi = folium.Map(location=[latitude, longitude], zoom_start=12)


# add markers of neighborhoods to the map
for lat, lng, num in zip(df_FFRe['venue_lat'], df_FFRe['venue_lon'], df_FFRe.index):
    label = '{}'.format( df_FFRe.iloc[num].venue_name )
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color = 'brown' if (df_FFRe.iloc[num].venue_name == 'Burger King') else 'blue',
        fill=True,
        #fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_etobi)
#
for lat, lng, num in zip(df_CSe['venue_lat'], df_CSe['venue_lon'], df_CSe.index):
    label = '{}'.format( df_CSe.iloc[num].venue_name )
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color = 'red' if (df_CSe.iloc[num].venue_name == 'Tim Hortons') else 'green',
        fill=True,
        #fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_etobi)

map_etobi

# END